File: sdk/python/aip/manifest.py
"""Manifest builder for AIP agents"""
from uuid import uuid4
from .types import Manifest, AgentInfo, Capability, Endpoints, TrustConfig


class ManifestBuilder:
    def __init__(self) -> None:
        self._agent = AgentInfo(id=str(uuid4()), name="")
        self._capabilities: list[Capability] = []
        self._endpoints = Endpoints(aip="")
        self._auth: list[str] = []
        self._trust: TrustConfig | None = None

    def agent(self, name: str, **kwargs: str) -> "ManifestBuilder":
        self._agent.name = name
        for k, v in kwargs.items():
            setattr(self._agent, k, v)
        return self

    def agent_id(self, id: str) -> "ManifestBuilder":
        self._agent.id = id
        return self

    def capability(self, cap: Capability) -> "ManifestBuilder":
        self._capabilities.append(cap)
        return self

    def endpoints(self, aip: str, health: str = "") -> "ManifestBuilder":
        self._endpoints = Endpoints(aip=aip, health=health)
        return self

    def trust(self, trust: TrustConfig) -> "ManifestBuilder":
        self._trust = trust
        return self

    def build(self) -> Manifest:
        assert self._agent.name, "Agent name required"
        assert self._endpoints.aip, "AIP endpoint required"
        assert self._capabilities, "At least one capability required"
        return Manifest(
            aip="0.1", agent=self._agent, capabilities=self._capabilities,
            endpoints=self._endpoints, auth_schemes=self._auth, trust=self._trust,
        )

File: sdk/typescript/src/manifest.ts
/**
 * Manifest builder — fluent API for constructing AIP agent manifests
 */
import { randomUUID } from 'crypto';
import { Manifest, AgentInfo, Capability, Endpoints, AuthConfig, TrustConfig } from './types';

export class ManifestBuilder {
  private manifest: Manifest;

  constructor() {
    this.manifest = {
      aip: '0.1',
      agent: { id: randomUUID(), name: '' },
      capabilities: [],
      endpoints: { aip: '' },
    };
  }

  agent(info: Partial<AgentInfo> & { name: string }): this {
    this.manifest.agent = { id: this.manifest.agent.id, ...info };
    return this;
  }

  capability(cap: Capability): this {
    this.manifest.capabilities.push(cap);
    return this;
  }

  endpoints(ep: Endpoints): this {
    this.manifest.endpoints = ep;
    return this;
  }

  auth(auth: AuthConfig): this {
    this.manifest.auth = auth;
    return this;
  }

  trust(trust: TrustConfig): this {
    this.manifest.trust = trust;
    return this;
  }

  build(): Manifest {
    if (!this.manifest.agent.name) throw new Error('Agent name is required');
    if (!this.manifest.endpoints.aip) throw new Error('AIP endpoint is required');
    if (this.manifest.capabilities.length === 0) throw new Error('At least one capability is required');
    return structuredClone(this.manifest);
  }
}

/** Validate a manifest object has required fields */
export function validateManifest(m: unknown): m is Manifest {
  const obj = m as Manifest;
  return (
    typeof obj?.aip === 'string' &&
    typeof obj?.agent?.id === 'string' &&
    typeof obj?.agent?.name === 'string' &&
    Array.isArray(obj?.capabilities) &&
    obj.capabilities.length > 0 &&
    typeof obj?.endpoints?.aip === 'string'
  );
}
